numpy/core/src/multiarray/ctors.c

```diff
@@ -1481,6 +1498,89 @@ _array_from_buffer_3118(PyObject *memoryview)
 
 }
 
+
+/*
+ * Attempts to extract an array from an array-like object.
+ *
+ * array-like is defined as either
+ *
+ * * an object implementing the PEP 3118 buffer interface;
+ * * an object with __array_struct__ or __array_interface__ attributes;
+ * * an object with an __array__ function.
+ *
+ * Returns Py_NotImplemented if a given object is not array-like;
+ * PyArrayObject* in case of success and NULL in case of failure.
```

Comment:
Worth noting that if it returns `Py_NotImplemented` the caller should _not_ decref it.

Reply:
Is this a reasonable thing to do (i.e. is it common within NumPy internals)? I could instead just incref `Py_NotImplemented`.

Reply:
incref ing would be better

Reply:
Done.